Bulk update

tanyaxp opened the pull request.

potsui committed:
REMOVE ON MERGE Run celery on devstack

Apply this commit to run celery on devstack

Make bulk update problem settings basic functionality

Currently, takes input values from user, `max_attempts` and
`show_answer`, and bulk updates them for existing problems in one go.

In the future, we need to implement the following:
- Option for instructor to update only one of those settings
- Catch exceptions
- Fix status polling (on error, reset, reload)
- Consider removing feedback for the submit and validate stages, since
  these take a negligible amount of time. This would also simplify
  fixing status polling
- Load default values from advanced settings or remove reference
- Move files to `openedx/stanford` and remove dependencies on `cms`
- Add styling and clean up existing stylesheet
- Unit tests
- Clean up all files for consistent syntax and style guide conformity
- Documentation

Allow option to update only one setting

Previously, we did not have an option for instructors to update only one
of the settings. This features adds checkboxes for instructors to
indicate what they would like to update, and implements data validation
in both browser and server to respond appropriately. We chose this
approach as opposed to having separate update buttons for each setting,
to reduce the amount of update time for when instructors would like to
change more than one setting simulatneously.

Fix proper status polling and UI feedback

Remove references to advanced settings

Previously, we had a skeleton for when we would load the default values
from advanced settings. We removed all references to advanced settings
in this skeleton and opted to simply load `max_attempts=0` and
`show_answer=always` as the default values shown.

Create reset button on error and success

Previously, the feedback list on errors would immediately reset and not
display any error messages to the user. The reset button allows users to
explicitly reset the internal state of the update process after they
have seen the error (or success) message and hide the feedback list for
a fresh UI again.

Fix feedback list, button, and input on resume

Previously, refreshing the page would display the wrong state and
button, and reset the user's previous input.  We now load the correct
update state in the feedback list, button, and user's previous input.

Clean up Python file for bulk update problems

Remove unnecessary imports, create more robust error catching, and
return more consistent error responses.

Fix jshint errors and reorder declarations

Delay showing reset button on complete

Fix python file issues exposed by unit tests

Changes:
- Import HttpResponseNotFound
- Separate try/except blocks for retrieving session_status_string to
  prevent use before declaration.

WIP Write unit tests

Break down into separate internal handlers

Use celery task for update

WIP Change to celery task

Moved code to separate tasks.py file to run bulk update as celery task.
Currently is able to update for small courses but worker is killed for
large courses because of memory consumption.

WIP Clean up python files

Remove duplicated and unused code, refactor

Fix memory issue and clean up files

Add bulk_operations to use single transaction in order to fix memory
issue. Change UI after feedback. Clean up code quality issues.

Modify advanced settings on bulk update

Per f2f discussion we determined that instructors in the general use
case would want to update both existing problems (already implemented)
and set the default for future problems in advanced settings (this
commit).

Clean up template and stylesheet

Add styling to UI and clean up code quality issues and redundancies in
stylesheets.

Add extra line for code quality

Republish problems if already published

Previously, bulk update would simply update problem settings as drafts
and leave the publish operation to the user within the course outline.
This was time consuming and inefficient. Now, we automatically republish
problems that have already been published. Problems that have have never
yet been published will remain as drafts.

Update UI to display changes

We add "updating advanced settings" as a distinct stage and reword the
interface to alert the user that published problems will auto publish.

Remove devstack debugging

Send email on celery bulk update task completion

Notifies the request user if their bulk update has failed or completed
successfully.

Change listeners for form input elements

Remove use of deprecated `bind()` and add listeners so that checkbox
will automatically get checked on input field change.

Add warning and link to course outline

Currently, we publish all public problems when they get bulk updated. It
may be that the problem already is a draft with other changes. We warn
the user to visit the Course Outline to ensure they want these changes
to also be made public before they perform a bulk update.

Remove accidental duplicate line

Break up long line into two

WIP Write unit tests

Respond to stv's feedback front-end

Respond to stv - tests

Respond to stv's feedback backend

Replace logging with celery logger

Move python files to openedx

Update test file due to feedback changes

Respond to f2f code review session

Clean up style issues, remove polling functionality, add more robust
data validation and error catching, etc.

Write basic suite of unit tests

Test handler responses and that advanced settings and problem settings
update. Does not yet test edge cases, email, etc.

Respond stv's feedback 2

Add unit tests and clean up test file

Fix pylint issues

Move static and template files from cms to openedx/stanford/cms/

Move static and template files from cms to openedx/stanford/cms/ to keep
our changes separate from upstream changes.

Remove scss file and apply css from upstream libraries

Apply css from upstream libraries

stvstnfrd review changes

REMOVE ON MERGE Run celery on devstack

Apply this commit to run celery on devstack

Make bulk update problem settings basic functionality

Currently, takes input values from user, `max_attempts` and
`show_answer`, and bulk updates them for existing problems in one go.

In the future, we need to implement the following:
- Option for instructor to update only one of those settings
- Catch exceptions
- Fix status polling (on error, reset, reload)
- Consider removing feedback for the submit and validate stages, since
  these take a negligible amount of time. This would also simplify
  fixing status polling
- Load default values from advanced settings or remove reference
- Move files to `openedx/stanford` and remove dependencies on `cms`
- Add styling and clean up existing stylesheet
- Unit tests
- Clean up all files for consistent syntax and style guide conformity
- Documentation

Allow option to update only one setting

Previously, we did not have an option for instructors to update only one
of the settings. This features adds checkboxes for instructors to
indicate what they would like to update, and implements data validation
in both browser and server to respond appropriately. We chose this
approach as opposed to having separate update buttons for each setting,
to reduce the amount of update time for when instructors would like to
change more than one setting simulatneously.

Fix proper status polling and UI feedback

Remove references to advanced settings

Previously, we had a skeleton for when we would load the default values
from advanced settings. We removed all references to advanced settings
in this skeleton and opted to simply load `max_attempts=0` and
`show_answer=always` as the default values shown.

Create reset button on error and success

Previously, the feedback list on errors would immediately reset and not
display any error messages to the user. The reset button allows users to
explicitly reset the internal state of the update process after they
have seen the error (or success) message and hide the feedback list for
a fresh UI again.

Fix feedback list, button, and input on resume

Previously, refreshing the page would display the wrong state and
button, and reset the user's previous input.  We now load the correct
update state in the feedback list, button, and user's previous input.

Clean up Python file for bulk update problems

Remove unnecessary imports, create more robust error catching, and
return more consistent error responses.

Fix jshint errors and reorder declarations

Delay showing reset button on complete

Fix python file issues exposed by unit tests

Changes:
- Import HttpResponseNotFound
- Separate try/except blocks for retrieving session_status_string to
  prevent use before declaration.

WIP Write unit tests

Break down into separate internal handlers

Use celery task for update

WIP Change to celery task

Moved code to separate tasks.py file to run bulk update as celery task.
Currently is able to update for small courses but worker is killed for
large courses because of memory consumption.

WIP Clean up python files

Remove duplicated and unused code, refactor

Fix memory issue and clean up files

Add bulk_operations to use single transaction in order to fix memory
issue. Change UI after feedback. Clean up code quality issues.

Modify advanced settings on bulk update

Per f2f discussion we determined that instructors in the general use
case would want to update both existing problems (already implemented)
and set the default for future problems in advanced settings (this
commit).

Clean up template and stylesheet

Add styling to UI and clean up code quality issues and redundancies in
stylesheets.

Add extra line for code quality

Republish problems if already published

Previously, bulk update would simply update problem settings as drafts
and leave the publish operation to the user within the course outline.
This was time consuming and inefficient. Now, we automatically republish
problems that have already been published. Problems that have have never
yet been published will remain as drafts.

Update UI to display changes

We add "updating advanced settings" as a distinct stage and reword the
interface to alert the user that published problems will auto publish.

Remove devstack debugging

Send email on celery bulk update task completion

Notifies the request user if their bulk update has failed or completed
successfully.

Change listeners for form input elements

Remove use of deprecated `bind()` and add listeners so that checkbox
will automatically get checked on input field change.

Add warning and link to course outline

Currently, we publish all public problems when they get bulk updated. It
may be that the problem already is a draft with other changes. We warn
the user to visit the Course Outline to ensure they want these changes
to also be made public before they perform a bulk update.

Remove accidental duplicate line

Break up long line into two

WIP Write unit tests

Respond to stv's feedback front-end

Respond to stv - tests

Respond to stv's feedback backend

Replace logging with celery logger

Move python files to openedx

Update test file due to feedback changes

Respond to f2f code review session

Clean up style issues, remove polling functionality, add more robust
data validation and error catching, etc.

Write basic suite of unit tests

Test handler responses and that advanced settings and problem settings
update. Does not yet test edge cases, email, etc.

Respond stv's feedback 2

Add unit tests and clean up test file

Fix pylint issues

Move static and template files from cms to openedx/stanford/cms/

Move static and template files from cms to openedx/stanford/cms/ to keep
our changes separate from upstream changes.

Remove scss file and apply css from upstream libraries

Apply css from upstream libraries

stvstnfrd review changes

Move bulkupdate URLs to stanford subdirectory

Commit:
Cache courses list on sysadmin courses page

We were performing this lookup multiple times in the POST handler;
this should increase speed and memory use, which has been problematic
with this feature.

DawoudSheraz committed:
Fix issue where refreshing page after adjusting video speed would result in video playing at 1.0x (regardless of speed selected) even though UI would show selected speed.

Issue fixed by cherry-picking upstream bugfix here: https://github.com/edx/edx-platform/pull/19888/commits/a5fe66863eef403ffa4606a22c81ec2a367f1986

Commit:
Remove unused 'plot' references in sysadmin

This appears to have been included with the origin commit that
introduced the sysadmin page [1].
It doesn't appear to have been ever used and, more importantly,
the view isn't passing it in the context to the template,
so it should be good to go.

- [1] ccc87337d63b593cad06cf4bc812c5bcf01b8be3

Commit:
Remove course listing from user tab of sysadmin

I'm admittedly unclear why we decided to include a listing of _courses_
in the _users_ tab, when this listing is already available via the
_courses_ tab, which is where you'd expect to find it.

Particularly problematic here is the impact this has on page loads..
For instances with a large number of courses [1], pulling and rendering
adds a nontrivial amount of overhead and leads to longer load times.

In our case, this often leads to a page timeout. Since this is the main
landing page for the sysadmin dashboard, it presents the appearance that
the entire dashboard has crashed. In not being able to access the main
page, you can't access any of the other tabs via navigation, unless you
manually enter the URL :\

[1] ~750 courses, in one example

Commit:
Remove unused version from footer

I'm not seeing this value ever initialized in `settings`,
or referenced anywhere else in platform, for that matter..

stvstnfrd commented:
@tanyaxp Here are the "${PATH}"-like variables I mentioned today.

Additional template paths (*.html), including wildcards, are added here:
- https://github.com/Stanford-Online/edx-platform/blob/master/openedx/stanford/cms/envs/common.py#L57-L59
- https://github.com/Stanford-Online/edx-platform/blob/master/openedx/stanford/cms/envs/common.py#L68-L70

And then static content (*.js, *.css, *.scss) here:
- https://github.com/Stanford-Online/edx-platform/blob/master/openedx/stanford/cms/envs/common.py#L65-L67

So I think we can group all of these files together within the djangoapp itself at:
- `openedx/stanford/cms/djangoapps/contentstore/templates/`
  - for the `*.html` files
- `openedx/stanford/cms/djangoapps/contentstore/static/`
  - for the static files

and the system _should_ locate them automatically, all the same 

🤞

stanford-online-robot committed:
Nominate release candidate

* origin/master:
  try create cache dir
  Skip gitlog pagination test
  Fix mongo query count failure
  Skip translation-related failure
  Stop installing mysql-server for tests
  Edit test value to be valid URL format
  Bump recommender
  Update circleci to v2.0
  Sort discovery search results by start date
  Add missing import
  Handle out-of-memory exception on Sysadmin Courses

stanford-online-robot committed:
Release to production

* origin/rc:
  try create cache dir
  Skip gitlog pagination test
  Fix mongo query count failure
  Skip translation-related failure
  Stop installing mysql-server for tests
  Edit test value to be valid URL format
  Bump recommender
  Update circleci to v2.0
  Sort discovery search results by start date
  Add missing import
  Handle out-of-memory exception on Sysadmin Courses

stanford-online-robot committed:
Backport to master

* origin/release:

stvstnfrd commented on cms/static/sass/_build-v1.scss:
```diff
@@ -95,4 +95,3 @@
 @import 'edx-pattern-library-shims/buttons';
 
 @import 'edx-pattern-library-shims/base/variables';
-
```
If we restore this line, we won't be making any fork to the file :)

stvstnfrd commented on cms/djangoapps/contentstore/views/__init__.py:
```diff
@@ -8,6 +8,7 @@
 from .utility import *
 from .utilities.captions import *
 from .utilities.bulksettings import *
+from openedx.stanford.cms.djangoapps.contentstore.views.utilities.bulkupdate import *
```
I _think_ we can remove this line;
I don't believe we're actively using it here, at this point.

Let's give it a shot..

tanyaxp replied:
This did not work

stvstnfrd reviewed (commented):
&nbsp;

Commit:
Merge pull request #884

* stv/sysadmin/cache-courses-origin:
  Cache courses list on sysadmin courses page

Commit:
Merge pull request #885

* stv/sysadmin/plots-origin:
  Remove unused 'plot' references in sysadmin

Commit:
Merge pull request #886

* stv/sysadmin/user-courses-origin:
  Remove course listing from user tab of sysadmin

Commit:
Remove empty file

It looks look like killed the contents of this directory,
but missed the __init__.py

Commit:
Merge pull request #887

* stv/kill/empty-file:
  Remove empty file

stanford-online-robot committed:
Nominate release candidate

* origin/master:
  Remove empty file
  Remove course listing from user tab of sysadmin
  Remove unused 'plot' references in sysadmin
  Cache courses list on sysadmin courses page

stanford-online-robot committed:
Release to production

* origin/rc:
  Remove empty file
  Remove course listing from user tab of sysadmin
  Remove unused 'plot' references in sysadmin
  Cache courses list on sysadmin courses page

stanford-online-robot committed:
Backport to master

* origin/release:

Commit:
Move Studio Bulk Utility tools to Stanford dirs

Commit:
Merge pull request #890

* stv/sysadmin/version:
  Remove unused version from footer

Commit:
Merge pull request #889

* stv/fork/bulk-studio:
  Move Studio Bulk Utility tools to Stanford dirs

stanford-online-robot committed:
Nominate release candidate

* origin/master:
  Remove unused version from footer
  Move Studio Bulk Utility tools to Stanford dirs

tanyaxp commented:
@stvstnfrd , @caesar2164 , I have pushed my changes, please review. I still need to works on tests when I am back on Monday.

stanford-online-robot committed:
Release to production

* origin/rc:
  Remove unused version from footer
  Move Studio Bulk Utility tools to Stanford dirs

stanford-online-robot committed:
Backport to master

* origin/release:

Commit:
Merge pull request #891

* ar/fix-video-speed-adjustments:
  Fix issue where refreshing page after adjusting video speed would result in video playing at 1.0x (regardless of speed selected) even though UI would show selected speed.

caesar2164 commented on lms/envs/devstack.py:
```diff
@@ -17,7 +17,8 @@
 # By default don't use a worker, execute tasks as if they were local functions
-CELERY_ALWAYS_EAGER = True
+CELERY_ALWAYS_EAGER = False
+FEATURES['ENABLE_SERVICE_STATUS'] = True
```
Is this from testing? Doesn't seem useful for this feature.

caesar2164 replied:
Also, another lms file being being modified.

caesar2164 commented on cms/envs/devstack.py:
```diff
@@ -62,6 +62,7 @@
 
 # By default don't use a worker, execute tasks as if they were local functions
 CELERY_ALWAYS_EAGER = True
+FEATURES['ENABLE_SERVICE_STATUS'] = True
```
Is this from testing? Doesn't seem useful for this feature.

caesar2164 commented on cms/templates/emails/utilities_bulk_update_done_message.txt:
```diff
@@ -0,0 +1,15 @@
+${_("The following settings were modified:")}
+
+% for key, value in modified_settings.iteritems():
+${_("{key}: {value}").format(key=key, value=value)}
```
No need to translate this string as there isn;t anything to actually translate...

caesar2164 commented on lms/djangoapps/dashboard/sysadmin.py:
```diff
@@ -300,15 +300,6 @@ def make_common_context(self):
-            self.msg += u'<li>{0} ({1})</li>'.format(
-                escape(course.id.to_deprecated_string()), course.location.to_deprecated_string())
-        self.msg += u'</ol>'
-
```
Why are you modifying an lms file?

caesar2164 commented on lms/templates/sysadmin_dashboard.html:
```diff
@@ -244,28 +242,6 @@ <h2>${datatable['title']}</h2>
-    %endfor
-
-%endif
-
```
LMS file again, shouldn't be in the PR.

caesar2164 reviewed (changes requested):
Quick once over.

It looks like some of the cruft might be from when you merged rather than rebased @tanyaxp.

@stvstnfrd you should probably help @tanyaxp clean up her branch. (Remove the merged commits from master, squash the commits from Po down a bit, and a general code cleanup?)

stvstnfrd commented on openedx/stanford/cms/djangoapps/contentstore/tests/test_bulkupdate.py:
```diff
@@ -0,0 +1,184 @@
+        store = modulestore()
+        problems = store.get_items(
+            self.course.id,
+            qualifiers={"category": 'problem'},
```
Nitpick: single quotes for consistency.

stvstnfrd commented on openedx/stanford/cms/djangoapps/contentstore/tests/test_bulkupdate.py:
```diff
@@ -0,0 +1,184 @@
+        for option in SHOW_ANSWER_OPTIONS[1:]:
+            self.assertContains(response, '<option value="{}">{}</option>'.format(option, option))
+
+    def test_bulkupdate_put_unsupported(self):
```
We should combine all of the unsupported HTTP requests into a single, DDT-driven test.

stvstnfrd commented on openedx/stanford/cms/djangoapps/contentstore/tests/test_bulkupdate.py:
```diff
@@ -0,0 +1,184 @@
+
+        problems = store.get_items(
+            self.course.id,
+            qualifiers={"category": 'problem'},
```
Nit: single quotes.

stvstnfrd commented on openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py:
```diff
@@ -0,0 +1,149 @@
+
+from cms.djangoapps.contentstore.utils import reverse_course_url
+
+from .tasks import bulk_update_problem_settings
```
All edx-platform imports should be grouped together.

stvstnfrd commented on openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py:
```diff
@@ -0,0 +1,149 @@
+        course_key_string
+    )
+    course_outline_url = reverse_course_url(
+        "course_handler",
```
Nit: single quotes.

stvstnfrd reviewed (changes requested):
@tanyaxp Here are some thoughts. Let's review this review in-person before going any further.

tanyaxp committed:
Partial update to code (re: review from Stv)

tanyaxp committed:
Javascript changes for Steven's feedback
Import changes for test

stvstnfrd commented:
@tanyaxp  Feel free to close this :)
